subsys/pm/device_runtime.c

```diff
@@ -138,6 +138,12 @@ int pm_device_runtime_get(const struct device *dev)
 		if (ret != 0) {
 			goto unlock;
 		}
+		/* Check if powering up this device failed */
+		if (atomic_test_bit(&pm->flags, PM_DEVICE_FLAG_TURN_ON_FAILED)) {
+			(void)pm_device_runtime_put(PM_DOMAIN(pm));
+			ret = -EAGAIN;
+			goto unlock;
+		}
```

Comment:
the code above calls `pm_device_runtime_get` again to wake-up the domain, which never sets `PM_DEVICE_FLAG_TURN_ON_FAILED`, maybe I'm missing something here.

Reply:
I'm not sure what you mean here. `PM_DEVICE_FLAG_TURN_ON_FAILED` is set on individual devices, not the domain itself.
1. The domain is powered on, which may result in `PM_DEVICE_FLAG_TURN_ON_FAILED` being set on the device we care about.
2. If it is set, power off the domain again

Reply:
what I mean here is that `PM_DEVICE_FLAG_TURN_ON_FAILED` is never set by the pm runtime code, so a project using pm runtime exclusively will never see that bit being set, again maybe I'm missing something.

Reply:
The way this currently works is: `pm_device_runtime_get` is called first on the device, then on the power domain.
`pm->action_cb(dev, PM_DEVICE_ACTION_RESUME)` on the power domain is calling `pm_device_children_action_run` in the power domain transition, which is calling the `pm_device_action_run`.
https://github.com/zephyrproject-rtos/zephyr/blob/523cd60a77525d1b2a52923cdf0459913cf069a2/drivers/power_domain/power_domain_gpio.c#L43-L52

I wasn't really aware that Runtime PM wasn't implemented in terms of `pm_device_action_run`.
The above flow isn't really intuitive, but I don't think it can be pulled into Runtime PM because that is only concerned with the ACTIVE and SUSPENDED states.

Reply:
Which is another way of saying that I think any project using power domains cannot exclusively use Runtime PM.

Reply:
@gmarull that is related with the discussion we had when there was the device runtime overall. I don't think it is a problem add this logic to `pm_action_run`. Actually I think we should change device runtime to use that.

I don't think we are promoting that API doing it, we still encourage devices to use the runtime API. But features are getting more complex and it seems to me that is error prone having the driver callback called directly from here.

Reply:
the actions TURN_ON and TURN_OFF are kind of orthogonal to device runtime. I mean, these actions are basically notifications to devices under a domain telling them whether or not they are clock gated.

If I understood correctly the workflow here for a device A under a domain B would be:
pm_device_runtime_get(A)
  |-> pm_device_runtime_get(B)
    |-> pm->action_run(B, RESUME)
      |-> pm_device_action_run(A, TURN_ON)
  pm->action_run(A, RESUME)

Reply:
> the actions TURN_ON and TURN_OFF are kind of orthogonal to device runtime. I mean, these actions are basically notifications to devices under a domain telling them whether or not they are clock gated.

Kind of, device runtime shouldn't be trying to resume devices which fail `TURN_ON`, which implies it needs to be at least aware of those actions. They may be notifications for devices internal to an SoC, but external sensors do need to run complex logic in these callbacks (setup gpios, configure to sleep mode, etc).

> If I understood correctly the workflow here for a device A under a domain B would be:

Almost, I would include the `pm_device_children_action_run` call:
```
pm_device_runtime_get(A)
  |-> pm_device_runtime_get(B)
    |-> pm->action_run(B, RESUME)
      |->pm_device_children_action_run(B, TURN_ON)
        |-> pm_device_action_run(A, TURN_ON)
        |-> pm_device_action_run(X, TURN_ON)
        |-> pm_device_action_run(Y, TURN_ON)
  |-> pm->action_run(A, RESUME)
```

Reply:
> @gmarull that is related with the discussion we had when there was the device runtime overall. I don't think it is a problem add this logic to `pm_action_run`. Actually I think we should change device runtime to use that.
> 
> I don't think we are promoting that API doing it, we still encourage devices to use the runtime API. But features are getting more complex and it seems to me that is error prone having the driver callback called directly from here.

The problem is that runtime/system requirements are different, so performed actions can't be the same (I already explained it in the original discussion). Shared logic is of course okay, as soon as it's kept internal. `pm_device_action_run()` is really a shortcut to device's PM callback, with no functionality like reference counting. If runtime PM is preferred, we should be careful on what is exposed as a public API.

Reply:
I agree. My point is that, I don't see why runtime pm cannot be implemented on top of `pm_device_action_run`, this is a shortcut fo the device callback with a minimal required logic on top. Regarding what is exposed, we have to be careful indeed. Though for now, I prefer to have this API allowed to be used by power domains to notify devices than have more states to be handled in device runtime. There is not clean way to manage multiple states in device runtime API.